New Serverless Pattern - Amazon Cognito user management using Amazon API Gateway and AWS Lambda (#2333)
aws-samples/serverless-patterns: apigw-lambda-cognito-sam-java/README.md

## How it works

Please refer to the architecture diagram below:

![End to End Architecture](images/architecture.png)

Here's a breakdown of the steps:

1. **Amazon API Gateway**: Receives the HTTP POST request containing the request details in JSON format. There are 3 POST methods exposed for create user, confirm user and login user operations.

2. **AWS Lambda**: Triggered by the API Gateway, this function written in Java uses Amazon Cognito API to create, confirm and login user.

3. **Amazon Cognito**: Depending on the operation invoked, user is created, confirmed or logged in the Amazon Cognito user pool.

4. **Response**: The Lambda function processes the Amazon Cognito API response and sends it back to the user via the API Gateway.

## Testing

1. You can use [curl](https://curl.se/) or any other tool of your choice to send a HTTP POST create user request to the API. Make sure to replace `Endpoint` with the corresponding values from your `sam deploy --guided` output. Also replace all other values. The `email` should be a valid email id.

   ```bash

>    curl -d '{"firstName": "{YourFirstName}", "lastName": "{YourLastName}", "email": "{your-email@mail.com}", "password": "{Y0urP#ssword}"}' -H 'Content-Type: application/json' {CreateUserAPIGatewayEndpoin}

Reviewer: Please leave the password value empty and rather suggest to use a secure value

Author: Fixed

   ```

   The API returns a response as below: 

   ```json
   {"isSuccessful":true,"statusCode":200,"cognitoUserId":"84c8xxxx-xxxx-xxxx-xxx-xxxxx9846de1","isConfirmed":false}
   ```

> 2. You should get an email with a verification code in the provided email id. Use the varification code to invoke the confirm user POST API:

Reviewer: ```suggestion
2. You should receive an email with a verification code at your provided email. Use the verification code to invoke the confirm user POST API:
```

Author: Fixed.

   ```bash
   curl -d '{"email": "{your-email@mail.com}", "code": "{VerificationCode}"}' -H 'Content-Type: application/json' {ConfirmUserAPIGatewayEndpoint}
   ```

   The API returns a response as below: 

   ```json
   {"isSuccessful":true,"statusCode":200}
   ```

3. Now, you can call the login POST API. This API will return the token:
You should get an email with a verification code in the provided email id. Use the varification code to invoke the confirm user POST API:
   ```bash

>    curl -d '{"email": "{your-email@mail.com}", "password": "{Y0urP#ssword}"}' -H 'Content-Type: application/json' {LoginUserAPIGatewayEndpoint}

Reviewer: See above

Author: Fixed.
